### PR TITLE
Implement RSA verification

### PR DESCRIPTION
**Pull request by Amxx**
Fix LIB-1238

Work in progress. Based on https://github.com/adria0/SolRsaVerify

Tests:
- parse `SigVer15_186-3.rsp` from [here](https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/dss/186-2rsatestvectors.zip) and automatically run all test examples.
- also run two extra tests from https://github.com/adria0/SolRsaVerify

For now, only sha256 is supported. `@nobles/hash` provides all the function we need to generate the digest, and `SigVer15_186-3.rsp` contains more tests.

Do we want to support more?

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [x] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: contracts/utils/cryptography/RSA.sol
Comment (ernestognw) [inline contracts/utils/cryptography/RSA.sol:46]: I guess we reject length less than `0x40` because it wouldn't be secure. I wonder if `0x40` was arbitrarily chosen. If so, we need to evaluate it carefully, as far as I remember, RSA's security is `p * q` so a 512 bits signature is crackable in reasonable time.

Found this as a reference, but seems like 512 bits (0x40 bytes) signatures are pretty much broken.
https://github.com/tomrittervg/cloud-and-control/blob/master/gnfs-info/factoring-howto.txt

[RFC 3447](https://datatracker.ietf.org/doc/html/rfc8017) is from 2003 and was superseded by [RFC 8017](https://datatracker.ietf.org/doc/html/rfc8017), though, I couldn't find a recommendation for the mod length. Allegedly, 512 bits security was [first broken in 1999](https://crypto.stackexchange.com/a/3933), so my estimations say that we might increase this to `0x80` at least.

Needs discussion

Comment (ernestognw) [inline contracts/utils/cryptography/RSA.sol:46]: `length != sig.length` corresponds to the first step of [RFC 3447 in section 8.2.2](https://datatracker.ietf.org/doc/html/rfc8017#section-8.2.2)

> 1. Length checking: If the length of the signature S is not k octets, output "invalid signature" and stop.

**Review by ernestognw — COMMENTED** [review]
Managed to navigate through [RFC 8017](https://datatracker.ietf.org/doc/html/rfc8017) and mostly looks good.

File: contracts/utils/cryptography/RSA.sol
Comment (ernestognw) [reply contracts/utils/cryptography/RSA.sol]: Added comments explaining what each byte is

**Comment by ernestognw** [comment]
I took the freedom of renaming the functions since one that receives a `bytes32` digest is not exclusively `sha256` as far as I can tell. Theoretically, a user can sign with RSA using any digest they like.

> For now, only sha256 is supported. @nobles/hash provides all the function we need to generate the digest, and SigVer15_186-3.rsp contains more tests.
>
> Do we want to support more?

Honestly, I don't think so, we'd require to adjust the param checks according to the length of the digest since we're counting backward to get the digest and allow optional digest info parameters.

I'm marking this as ready for review since I think it's already good enough and I already feel comfortable with its security.

**Review by ernestognw — COMMENTED** [review]
In my opinion, I think the last thing to resolve is whether to restrict `n` to have a minimum length or not. I added a note (probably should be a warning) suggesting that no key with length less than 2048 bits should be used.

If we don't find a reason to continue allowing smaller lengths, then I'd advocate for restricting the function.

Aside from that, we need to cover the missing test cases (should be trivial) and it'll be good to go

File: .changeset/curvy-crabs-repeat.md
Comment (Amxx) [inline .changeset/curvy-crabs-repeat.md:5]:
```suggestion
`RSA`: Library to verify signatures according to RFC 8017 Signature Verification Operation
```

**Comment by ernestognw** [comment]
[Line 106](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4952/files#diff-d5604af0f028f31eba96d4f25df528d80de6aa9b98f04b9cfd2607ab3f0b9391R106) is pretty difficult to reach as it requires to set up an attack where we forge two messages such that once signed we can derive a third signature that decrypts wrongly to the message. (Note by "decrypt" I'm referring to the `modExp` operation to recover the message)

I think it's ok if we don't have a test case for that rn. In the future we should look for adding fuzzing, however, we don't have an easy way of fuzzing keys unless we use `--ffi`.

**Review by ernestognw — APPROVED** [review]
Looks solid and I feel comfortable with the result, I'll leave it to your consideration @Amxx.

File: contracts/utils/cryptography/RSA.sol
Comment (cairoeth) [inline contracts/utils/cryptography/RSA.sol:52]: do we have to check that buffer (`Math.tryModExp(s, e, n)`) = s to verify that the signature is valid for the public key (n, e)?

File: contracts/utils/cryptography/RSA.sol
Comment (cairoeth) [inline contracts/utils/cryptography/RSA.sol:108]: what do you think of how [this implementation](https://github.com/Legrandin/pycryptodome/blob/master/lib/Crypto/Signature/pkcs1_15.py#L133-L139) checks for this?


> By comparing the full encodings (as opposed to checking each
> of its components one at a time) we avoid attacks to the padding
> scheme like [Bleichenbacher's](https://medium.com/@c0D3M/bleichenbacher-attack-explained-bc630f88ff25)

File: contracts/utils/cryptography/RSA.sol
Comment (cairoeth) [inline contracts/utils/cryptography/RSA.sol:136]: because `_unsafeReadBytes2` is only used once, would it be cleaner if we just do the bytes2 casting [here](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4952/files#diff-d5604af0f028f31eba96d4f25df528d80de6aa9b98f04b9cfd2607ab3f0b9391R112)?

File: contracts/utils/cryptography/RSA.sol
Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:55]: tryModExp doesn't do that check, and we have nothing that guarantee that s < n.

We should probably implement a bytes comparaison function, and use it here.

File: contracts/utils/cryptography/RSA.sol
Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:52]: The result of `s ** e % n` is buffer. We don't check that it is equal to S. I don't think it should be equal to S.

We check that is it a properly formated "predicate" that contains the "digest", and all the other formating stuff.

File: contracts/utils/cryptography/RSA.sol
Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol]: I'm not sure what are the length of the objects being compared. This may not apply to solidity.

Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol]: If we were able to cap the size of the padding, we could keep track of all the possible hashes of string solely composed of `FF`, and compare the hash of that section ...

But I'm not really sure we can have a cap on paddingEnd (that requires a cap on length)

File: contracts/utils/cryptography/RSA.sol
Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:55]: we could use something like 

```
enum CmpResult { Eq, Lt, Gt }

function compare(bytes memory a, bytes memory b) pure returns (CmpResult) {
    unchecked {
        uint256 i = 0; // pointer on the a buffer
        uint256 j = 0; // pointer on the b buffer
        if (a.length > b.length) {
            while (i < a.length - b.length) {
                if (a[i++] != 0) return CmpResult.Gt;
            }
        } else if (a.length < b.length) {
            while (j < b.length - a.length) {
                if (b[j++] != 0) return CmpResult.Lt;
            }
        }
        while (i < a.length) { // eq j < b.length
            bytes1 ai = a[i++];
            bytes1 bj = b[j++];
            if (ai < bj) {
                return CmpResult.Lt;
            } else if (ai > bj) {
                return CmpResult.Gt;
            }
        }
        return CmpResult.Eq;
    }
}
```

Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:55]: @frangio 
I think we need this to avoid signature maleability ... but it also feels like an expensive check. WDYT?

We could write it in assembly ...
I'm going to check for other existing implementation

Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:55]: or we could check that 
```solidity
keccak256(s) == keccak256(modExp(s, 1, n))
```

Comment (ernestognw) [reply contracts/utils/cryptography/RSA.sol:55]: I'd be fine with `keccak256(s) == keccak256(modExp(s, 1, n))`.

`modExp` is only 200 gas if `e == 1`, and the cost of both keccak256 is 30.
The bottom line is that we're adding ~300 gas units, which is about 12 cents at 100gwei and ETH at 4000. 
I think this is good enough, and the loop might be even more costly if we end up accepting 4096-bit keys.

Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:55]: There is also the cost of the two keccak. I suspect the loop might be cheaper (particularly is we do some fancy assembly that checks in blocks of 32 bytes)

Comment (ernestognw) [reply contracts/utils/cryptography/RSA.sol:55]: Right, fair points. I also didn't consider under normal circumstances, the `compare` function finishes as early as possible.

If we're adding the `compare` function, I'd put it in `Math.sol`, wdyt?

Comment (ernestognw) [reply contracts/utils/cryptography/RSA.sol:55]: Note that we have already validated that `s.length == n.length` above, do you think the following would work?

```
enum CmpResult { Eq, Lt, Gt }
function compare(bytes memory a, bytes memory b) pure returns (CmpResult) {
    uint256 higherA = 0;
    uint256 higherB = 0;
    assembly ("memory-safe") {
        higherA := mload(add(a, 0x20))
        higherB := mload(add(b, 0x20))
    }
    return
        higherA == higherB
            ? CmpResult.Eq
            : higherA < higherB
                ? CmpResult.Lt
                : CmpResult.Gt;
}
```

or at least:

```
uint256 i = 0; // pointer on the a buffer
uint256 j = 0; // pointer on the b buffer
if (a.length > b.length) {
    while (i < a.length - b.length) {
        if (a[i++] != 0) return CmpResult.Gt;
    }
} else if (a.length < b.length) {
    while (j < b.length - a.length) {
        if (b[j++] != 0) return CmpResult.Lt;
    }
}
while (i < a.length) { // eq j < b.length
    bytes1 ai = a[i++];
    bytes1 bj = b[j++];
    if (ai < bj) {
        return CmpResult.Lt;
    } else if (ai > bj) {
        return CmpResult.Gt;
    }
}
return CmpResult.Eq;
```

Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:55]: first option only checks the first 32 bytes ... if the difference is further its not catching it. Arbitrary length means we need some loop somewhere.

Comment (ernestognw) [reply contracts/utils/cryptography/RSA.sol:55]: worst case in the first option is that they're 32-bytes equal, would it be possible to just require an explicit `Gt` and revert otherwise?

Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:67]: https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4952/commits/abe598d94f12e51648b9ea0e741083043e834e92

File: contracts/utils/cryptography/RSA.sol
Comment (ernestognw) [inline contracts/utils/cryptography/RSA.sol:76]: Given the verification algorithm considers optional algorithm parameters, then it's dangerous to protect for replay attacks by using a hash even if we make sure `s` is less than `n`. We should warn about that in the docs.

File: contracts/utils/cryptography/RSA.sol
Comment (frangio) [reply contracts/utils/cryptography/RSA.sol:67]: Do we know what would be the average cost of this check?

File: contracts/utils/cryptography/RSA.sol
Comment (frangio) [inline contracts/utils/cryptography/RSA.sol:55]:
```suggestion
                uint256 p = Math.min(i, length - 0x20);
                bytes32 si = _unsafeReadBytes32(s, p);
                bytes32 ni = _unsafeReadBytes32(n, p);
```

This code reverts if length is < 0x20... but I don't think that's realistic is it?

File: contracts/utils/cryptography/RSA.sol
Comment (ernestognw) [inline contracts/utils/cryptography/RSA.sol:59]: Shall we mention this?

```suggestion
                // if (si > ni) return false; // Assuming an honest input, this check makes it more expensive for the caller
                if (si < ni) {
                    ok = true;
                    break;
                }
```

File: contracts/utils/cryptography/RSA.sol
Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol:67]: The for could be long (64 iterations for a 4096 key)  ... but there is a break !

All real cases should have a difference somewhere in the first 32bytes (having such a large equality is not realisitc IMO).

One iteration should be way under 100 gas.

File: contracts/utils/cryptography/RSA.sol
Comment (ernestognw) [reply contracts/utils/cryptography/RSA.sol:59]: We know `length >= 0x40` from a check above

File: contracts/utils/cryptography/RSA.sol
Comment (ernestognw) [reply contracts/utils/cryptography/RSA.sol:55]: We know `length >= 0x40` from a check above

File: contracts/utils/cryptography/RSA.sol
Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol]: I don't understand the comment.

Comment (ernestognw) [reply contracts/utils/cryptography/RSA.sol]: If `si < ni`, then it's not needed to check the next memory word (32 bytes).  However, this only optimizes for an attacker, and wouldn't be needed for an honest caller.

I just found it weird that it loops until the end.

Comment (Amxx) [reply contracts/utils/cryptography/RSA.sol]: actually we had a bug here

File: contracts/utils/cryptography/RSA.sol
Comment (frangio) [reply contracts/utils/cryptography/RSA.sol]: Ok nice! In that case this can be unchecked.

File: contracts/utils/cryptography/RSA.sol
Comment (cairoeth) [inline contracts/utils/cryptography/RSA.sol:61]: if the loop breaks as soon as it finds one segment where sp < np (say the first 32-byte segment), wouldn't it overlook segments where sp > np later in the array (say the second 32-byte segment), potentially leading to incorrect conclusions about the overall order of s and n?

Comment (cairoeth) [reply contracts/utils/cryptography/RSA.sol:61]: maybe im not understand the assumption of `s < n globally`

File: test/utils/cryptography/RSA.test.js
Comment (cairoeth) [inline test/utils/cryptography/RSA.test.js:104]: this test is actually not reverting at the modExp step, but instead [here](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4952/files#diff-d5604af0f028f31eba96d4f25df528d80de6aa9b98f04b9cfd2607ab3f0b9391R62) because of `sp > np`. is there a way we can have more accurate checks when the return data is a boolean? maybe by measuring gas?

File: test/utils/cryptography/RSA.test.js
Comment (cairoeth) [inline test/utils/cryptography/RSA.test.js:98]: if module is 0, then it will revert at the s < n check, so maybe we can force the precompile to always revert for this case:
```suggestion
      setCode("0x0500000000000000000000000000000000000000", "0x60006000fd");
      const data = ethers.toUtf8Bytes('hello world!');
```

File: contracts/utils/cryptography/RSA.sol
Comment (frangio) [reply contracts/utils/cryptography/RSA.sol]: >64 iterations for a 4096 key

Are you sure about this? Isn't it 4096 bits, so 512 bytes, which is 16 32-byte words.

**Review by ernestognw — DISMISSED** [review]
Added some docs as well. LGTM.

I'd wait for a second approval.

**Review by cairoeth — APPROVED** [review]
lgtm -- @ernestognw: I added a replay protection notice on the docs page

File: docs/modules/ROOT/pages/utilities.adoc
Comment (ernestognw) [inline docs/modules/ROOT/pages/utilities.adoc:57]:
```suggestion
IMPORTANT: Always use keys of at least 2048 bits. Additionally, be aware that PKCS#1 v1.5 allows for replayability due to the possibility of arbitrary optional parameters added to the message construction before signing it. To prevent replay attacks, consider including an onchain nonce or unique identifier in the message.
```
```suggestion
IMPORTANT: Always use keys of at least 2048 bits. Additionally, be aware that PKCS#1 v1.5 allows for replayability due to the possibility of arbitrary optional parameters. To prevent replay attacks, consider including an onchain nonce or unique identifier in the message.
```
```suggestion
IMPORTANT: Always use keys of at least 2048 bits. Additionally, be aware that PKCS#1 v1.5 allows for replayability due to the possibility of arbitrary optional parameters. To prevent replay attacks, consider including an onchain nonce or unique identifier in the message.
```

**Comment by ernestognw** [comment]
> lgtm -- @ernestognw: I added a replay protection notice on the docs page

Thanks, looking good. Approving again

**Review by ernestognw — APPROVED** [review]

**Pull request merged**
